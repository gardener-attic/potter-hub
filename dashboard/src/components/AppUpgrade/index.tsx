import AppUpgrade from "./AppUpgrade";

export default AppUpgrade;
