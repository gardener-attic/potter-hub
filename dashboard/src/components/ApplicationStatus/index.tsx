import ApplicationStatus from "./ApplicationStatus";

export default ApplicationStatus;
