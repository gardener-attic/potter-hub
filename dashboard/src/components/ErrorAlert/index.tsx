import ErrorAlertHeader from "./ErrorAlertHeader";
import ErrorSelector from "./ErrorSelector";
import MessageAlert from "./MessageAlert";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import ServiceBrokersNotFoundAlert from "./ServiceBrokersNotFoundAlert";
import ServiceCatalogNotInstalledAlert from "./ServiceCatalogNotInstalledAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export {
  ErrorAlertHeader,
  ErrorSelector,
  MessageAlert,
  NotFoundErrorAlert,
  PermissionsErrorAlert,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
  UnexpectedErrorAlert,
};
