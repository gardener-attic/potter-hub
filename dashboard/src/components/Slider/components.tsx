// CODE FROM: https://sghall.github.io/react-compound-slider/#/slider-demos/horizontal
import * as React from "react";
import { GetHandleProps, GetTrackProps, SliderItem } from "react-compound-slider";

// *******************************************************
// HANDLE COMPONENT
// *******************************************************
interface IHandleProps {
  domain: number[];
  handle: SliderItem;
  getHandleProps: GetHandleProps;
}

export const Handle: React.SFC<IHandleProps> = ({
  domain: [min, max],
  handle: { id, value, percent },
  getHandleProps,
}) => (
  <div
    role="slider"
    aria-valuemin={min}
    aria-valuemax={max}
    aria-valuenow={value}
    style={{
      left: `${percent}%`,
      position: "absolute",
      marginLeft: "-11px",
      marginTop: "-6px",
      zIndex: 2,
      width: 24,
      height: 24,
      cursor: "pointer",
      borderRadius: "50%",
      boxShadow: "1px 1px 1px 1px rgba(0, 0, 0, 0.2)",
      backgroundColor: "#34568f",
    }}
    {...getHandleProps(id)}
  />
);

// *******************************************************
// TRACK COMPONENT
// *******************************************************
interface ITrackProps {
  source: SliderItem;
  target: SliderItem;
  getTrackProps: GetTrackProps;
}

export const Track: React.SFC<ITrackProps> = ({ source, target, getTrackProps }) => (
  <div
    style={{
      position: "absolute",
      height: 14,
      zIndex: 1,
      backgroundColor: "#7aa0c4",
      borderRadius: 7,
      cursor: "pointer",
      left: `${source.percent}%`,
      width: `${target.percent - source.percent}%`,
    }}
    {...getTrackProps()}
  />
);
