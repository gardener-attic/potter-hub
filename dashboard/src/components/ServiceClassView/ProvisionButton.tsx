import { RouterAction } from "connected-react-router";
import * as React from "react";
import Modal from "react-modal";

import { IClusterServiceClass } from "../../shared/ClusterServiceClass";
import { IServicePlan } from "../../shared/ServiceCatalog";
import { IRBACRole } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import SchemaForm from "../SchemaForm";

import { JSONSchema6 } from "json-schema";
import { ISubmitEvent } from "react-jsonschema-form";

interface IProvisionButtonProps {
  namespace: string;
  error?: Error;
  selectedClass: IClusterServiceClass;
  selectedPlan: IServicePlan;
  provision: (
    releaseName: string,
    namespace: string,
    className: string,
    planName: string,
    parameters: {},
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
}

interface IProvisionButtonState {
  isProvisioning: boolean;
  modalIsOpen: boolean;
  name: string;
  displayNameForm: boolean;
}

const RequiredRBACRoles: IRBACRole[] = [
  {
    apiGroup: "servicecatalog.k8s.io",
    resource: "serviceinstances",
    verbs: ["create"],
  },
];

class ProvisionButton extends React.Component<IProvisionButtonProps, IProvisionButtonState> {
  public state: IProvisionButtonState = {
    displayNameForm: true,
    isProvisioning: false,
    modalIsOpen: false,
    name: "",
  };

  public render() {
    const { selectedClass, selectedPlan, error, namespace } = this.props;
    let schema = selectedPlan.spec.instanceCreateParameterSchema;
    if (!schema) {
      schema = {
        properties: {
          kubeappsRawParameters: {
            title: "Parameters",
            type: "object",
          },
        },
        type: "object",
      };
    }

    return (
      <div className="ProvisionButton">
        {this.state.isProvisioning && <div>Provisioning...</div>}
        <button
          className="button button-primary button-small"
          onClick={this.openModel}
          disabled={this.state.isProvisioning}
        >
          Provision
        </button>
        <Modal
          isOpen={this.state.modalIsOpen}
          onRequestClose={this.closeModal}
          contentLabel="Modal"
          className={this.state.displayNameForm ? "centered-modal" : ""}
        >
          {error && (
            <ErrorSelector
              error={error}
              resource={selectedClass.spec.externalName}
              action="provision"
              defaultRequiredRBACRoles={{ provision: RequiredRBACRoles }}
              namespace={namespace}
            />
          )}
          {this.state.displayNameForm ? (
            <SchemaForm schema={this.nameSchema()} onSubmit={this.handleNameChange}>
              <div>
                <button className="button button-primary" type="submit">
                  Continue
                </button>
                <button className="button" onClick={this.closeModal}>
                  Cancel
                </button>
              </div>
            </SchemaForm>
          ) : (
            <SchemaForm schema={schema} onSubmit={this.handleProvision}>
              <div>
                <button
                  className="button button-primary"
                  type="submit"
                  disabled={this.state.isProvisioning}
                >
                  Submit
                </button>
                <button
                  className="button"
                  onClick={this.handleBackButton}
                  disabled={this.state.isProvisioning}
                >
                  Back
                </button>
              </div>
            </SchemaForm>
          )}
        </Modal>
      </div>
    );
  }

  public openModel = () => {
    this.setState({
      modalIsOpen: true,
    });
  };

  public closeModal = () => {
    this.setState({
      modalIsOpen: false,
    });
  };

  public handleBackButton = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    this.setState({ displayNameForm: true });
  };

  public handleNameChange = ({ formData }: ISubmitEvent<{ Name: string }>) => {
    this.setState({ name: formData.Name, displayNameForm: false });
  };

  public handleProvision = async ({
    formData,
  }: ISubmitEvent<{ name: string; kubeappsRawParameters: {} }>) => {
    const { namespace, provision, push, selectedClass, selectedPlan } = this.props;
    const { name } = this.state;
    this.setState({ isProvisioning: true });

    const { kubeappsRawParameters, ...rest } = formData;
    if (selectedClass && selectedPlan) {
      const provisioned = await provision(
        name,
        namespace,
        selectedClass.spec.externalName,
        selectedPlan.spec.externalName,
        kubeappsRawParameters || rest,
      );
      if (provisioned) {
        push(
          `/services/brokers/${selectedClass.spec.clusterServiceBrokerName}/instances/ns/${namespace}/${name}`,
        );
      } else {
        this.setState({ isProvisioning: false });
      }
    }
  };

  private nameSchema(): JSONSchema6 {
    return {
      properties: {
        Name: {
          default: this.state.name,
          description: "Name for ServiceInstance",
          type: "string",
        },
      },
      required: ["Name"],
      type: "object",
    };
  }
}

export default ProvisionButton;
