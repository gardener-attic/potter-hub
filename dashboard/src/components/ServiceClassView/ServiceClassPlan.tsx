import { RouterAction } from "connected-react-router";
import * as React from "react";
import { IClusterServiceClass } from "shared/ClusterServiceClass";
import { IServicePlan } from "shared/ServiceCatalog";
import ProvisionButton from "./ProvisionButton";

interface IServiceClassPlanProps {
  svcClass: IClusterServiceClass;
  svcPlan: IServicePlan;
  createError?: Error;
  provision: (
    instanceName: string,
    namespace: string,
    className: string,
    planName: string,
    parameters: {},
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  namespace: string;
}

const ServiceClassPlan: React.SFC<IServiceClassPlanProps> = props => {
  const { svcClass, svcPlan, createError, provision, push, namespace } = props;
  const { spec } = svcPlan;
  const { externalMetadata } = spec;
  const name = externalMetadata ? externalMetadata.displayName : spec.externalName;
  const description =
    externalMetadata && externalMetadata.bullets ? externalMetadata.bullets : [spec.description];
  const bullets = (
    <div>
      <ul className="margin-reset">
        {description.map(bullet => (
          <li key={bullet}>{bullet}</li>
        ))}
      </ul>
    </div>
  );

  return (
    <tr key={svcPlan.spec.externalID}>
      <td>
        <b>{name}</b>
      </td>
      <td>{bullets}</td>
      <td className="text-r">
        <ProvisionButton
          selectedClass={svcClass}
          selectedPlan={svcPlan}
          provision={provision}
          push={push}
          namespace={namespace}
          error={createError}
        />
      </td>
    </tr>
  );
};

export default ServiceClassPlan;
