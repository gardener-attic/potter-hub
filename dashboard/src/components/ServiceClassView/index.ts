import ServiceClassView from "./ServiceClassView";

export default ServiceClassView;
