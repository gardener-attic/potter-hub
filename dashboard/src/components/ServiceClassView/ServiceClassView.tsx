import { RouterAction } from "connected-react-router";
import * as React from "react";

import LoadingWrapper from "../../components/LoadingWrapper";
import PageHeader from "../../components/PageHeader";
import { IServiceCatalogState } from "../../reducers/serviceCatalog";
import { IClusterServiceClass } from "../../shared/ClusterServiceClass";
import { IRBACRole } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import ServiceClassPlan from "./ServiceClassPlan";

const RequiredRBACRoles: IRBACRole[] = [
  {
    apiGroup: "servicecatalog.k8s.io",
    clusterWide: true,
    resource: "clusterserviceclasses",
    verbs: ["list"],
  },
  {
    apiGroup: "servicecatalog.k8s.io",
    clusterWide: true,
    resource: "clusterserviceplans",
    verbs: ["list"],
  },
];

interface IServiceClassViewProps {
  classes: IServiceCatalogState["classes"];
  classname: string;
  createError?: Error;
  error?: Error;
  getClasses: () => Promise<any>;
  getPlans: () => Promise<any>;
  plans: IServiceCatalogState["plans"];
  provision: (
    instanceName: string,
    namespace: string,
    className: string,
    planName: string,
    parameters: {},
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  namespace: string;
}

class ServiceClassView extends React.Component<IServiceClassViewProps> {
  public componentDidMount() {
    this.props.getClasses();
    this.props.getPlans();
  }

  public render() {
    const {
      createError,
      classes,
      classname,
      error,
      plans,
      provision,
      push,
      namespace,
    } = this.props;
    const loaded = !classes.isFetching && !plans.isFetching;
    let classPlans: JSX.Element[] = [];
    let svcClass: IClusterServiceClass | undefined;

    if (loaded) {
      svcClass =
        classes.list.find(potential => !!(potential.spec.externalName === classname)) || undefined;
      if (svcClass) {
        const foundSVCClass = svcClass;
        const filteredPlans = plans.list.filter(
          plan => plan.spec.clusterServiceClassRef.name === foundSVCClass.metadata.name,
        );
        classPlans = filteredPlans.map(plan => {
          return (
            <ServiceClassPlan
              key={plan.metadata.name}
              svcClass={foundSVCClass}
              svcPlan={plan}
              provision={provision}
              push={push}
              createError={createError}
              namespace={namespace}
            />
          );
        });
      }
    }

    return (
      <div className="container">
        <PageHeader>
          <h1>Plans: {classname}</h1>
        </PageHeader>
        <main>
          <LoadingWrapper loaded={loaded}>
            <div className="class-view">
              {error && (
                <ErrorSelector
                  error={error}
                  resource="Service Plans"
                  action="list"
                  defaultRequiredRBACRoles={{ list: RequiredRBACRoles }}
                />
              )}
              <div>
                <p>Service Plans available for provisioning under {classname}</p>
                <table className="striped">
                  <thead>
                    <tr>
                      <th>Name</th>
                      <th>Specs</th>
                      <th />
                    </tr>
                  </thead>
                  <tbody>{classPlans}</tbody>
                </table>
              </div>
            </div>
          </LoadingWrapper>
        </main>
      </div>
    );
  }
}

export default ServiceClassView;
