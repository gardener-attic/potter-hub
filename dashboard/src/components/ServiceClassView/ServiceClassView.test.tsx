import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import ServiceClassView from ".";
import { ErrorSelector } from "../../components/ErrorAlert";
import { IClusterServiceClass } from "../../shared/ClusterServiceClass";
import { IServicePlan } from "../../shared/ServiceCatalog";
import itBehavesLike from "../../shared/specs";
import { ForbiddenError } from "../../shared/types";
import ServiceClassPlan from "./ServiceClassPlan";

const defaultName = "my-class";
const defaultNS = "default";
const defaultProps = {
  classes: { isFetching: false, list: [] },
  classname: defaultName,
  getClasses: jest.fn(),
  getPlans: jest.fn(),
  plans: { isFetching: false, list: [] },
  provision: jest.fn(),
  push: jest.fn(),
  namespace: defaultNS,
};

context("while fetching components", () => {
  const props = { ...defaultProps, classes: { isFetching: true, list: [] } };

  itBehavesLike("aLoadingComponent", { component: ServiceClassView, props });

  it("matches the snapshot", () => {
    const wrapper = shallow(<ServiceClassView {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a Application header", () => {
    const wrapper = shallow(<ServiceClassView {...props} />);
    expect(wrapper.find("h1").text()).toContain(defaultName);
  });
});

context("when all the components are loaded", () => {
  it("shows an error when listing classes if it exists", () => {
    const wrapper = shallow(<ServiceClassView {...defaultProps} error={new ForbiddenError()} />);
    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(ErrorSelector).props()).toMatchObject({
      error: new ForbiddenError(),
    });
  });

  context("when an service class is available", () => {
    let classes: any = {};
    let plans: any = {};
    beforeEach(() => {
      classes = {
        isFetching: false,
        list: [
          {
            metadata: {
              name: defaultName,
              uid: `class-${defaultName}`,
            },
            spec: {
              bindable: true,
              externalName: defaultName,
              description: "this is a class",
              externalMetadata: {
                imageUrl: "img.png",
              },
            },
          } as IClusterServiceClass,
        ],
      };
      plans = {
        isFetching: false,
        list: [
          {
            metadata: {
              name: defaultName,
            },
            spec: {
              externalName: defaultName,
              externalID: `plan-${defaultName}`,
              description: "this is a plan",
              clusterServiceClassRef: {
                name: defaultName,
              },
              free: true,
            },
          } as IServicePlan,
        ],
      };
    });

    it("should show the class info", () => {
      const wrapper = shallow(
        <ServiceClassView {...defaultProps} classes={classes} plans={plans} />,
      );
      expect(wrapper.find(ServiceClassPlan)).toExist();
      expect(wrapper).toMatchSnapshot();
    });

    it("should avoid plans that doesn't belong to the class", () => {
      const newPlan = {
        metadata: {
          name: defaultName,
        },
        spec: {
          externalName: defaultName,
          externalID: `plan-${defaultName}`,
          description: "this is a plan",
          clusterServiceClassRef: {
            name: `not-${defaultName}`,
          },
          free: true,
        },
      } as IServicePlan;
      plans.list.push(newPlan);
      const wrapper = shallow(
        <ServiceClassView {...defaultProps} classes={classes} plans={plans} />,
      );
      const planList = wrapper.find(ServiceClassPlan);
      expect(planList).toExist();
      expect(planList.length).toBe(1);
    });
  });
});
