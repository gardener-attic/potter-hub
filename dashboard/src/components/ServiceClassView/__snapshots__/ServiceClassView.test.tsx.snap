// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when all the components are loaded when an service class is available should show the class info 1`] = `
<div
  className="container"
>
  <PageHeader>
    <h1>
      Plans: 
      my-class
    </h1>
  </PageHeader>
  <main>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <div
        className="class-view"
      >
        <div>
          <p>
            Service Plans available for provisioning under 
            my-class
          </p>
          <table
            className="striped"
          >
            <thead>
              <tr>
                <th>
                  Name
                </th>
                <th>
                  Specs
                </th>
                <th />
              </tr>
            </thead>
            <tbody>
              <ServiceClassPlan
                key="my-class"
                namespace="default"
                provision={[MockFunction]}
                push={[MockFunction]}
                svcClass={
                  Object {
                    "metadata": Object {
                      "name": "my-class",
                      "uid": "class-my-class",
                    },
                    "spec": Object {
                      "bindable": true,
                      "description": "this is a class",
                      "externalMetadata": Object {
                        "imageUrl": "img.png",
                      },
                      "externalName": "my-class",
                    },
                  }
                }
                svcPlan={
                  Object {
                    "metadata": Object {
                      "name": "my-class",
                    },
                    "spec": Object {
                      "clusterServiceClassRef": Object {
                        "name": "my-class",
                      },
                      "description": "this is a plan",
                      "externalID": "plan-my-class",
                      "externalName": "my-class",
                      "free": true,
                    },
                  }
                }
              />
            </tbody>
          </table>
        </div>
      </div>
    </LoadingWrapper>
  </main>
</div>
`;

exports[`while fetching components loading spinner matches the snapshot 1`] = `
<div
  className="container"
>
  <PageHeader>
    <h1>
      Plans: 
      my-class
    </h1>
  </PageHeader>
  <main>
    <LoadingWrapper
      loaded={false}
      type={0}
    >
      <div
        className="class-view"
      >
        <div>
          <p>
            Service Plans available for provisioning under 
            my-class
          </p>
          <table
            className="striped"
          >
            <thead>
              <tr>
                <th>
                  Name
                </th>
                <th>
                  Specs
                </th>
                <th />
              </tr>
            </thead>
            <tbody />
          </table>
        </div>
      </div>
    </LoadingWrapper>
  </main>
</div>
`;

exports[`while fetching components matches the snapshot 1`] = `
<div
  className="container"
>
  <PageHeader>
    <h1>
      Plans: 
      my-class
    </h1>
  </PageHeader>
  <main>
    <LoadingWrapper
      loaded={false}
      type={0}
    >
      <div
        className="class-view"
      >
        <div>
          <p>
            Service Plans available for provisioning under 
            my-class
          </p>
          <table
            className="striped"
          >
            <thead>
              <tr>
                <th>
                  Name
                </th>
                <th>
                  Specs
                </th>
                <th />
              </tr>
            </thead>
            <tbody />
          </table>
        </div>
      </div>
    </LoadingWrapper>
  </main>
</div>
`;
